ark/phenotyping/cell_cluster_utils.py

```diff
@@ -604,58 +604,139 @@ def cell_consensus_cluster(fovs, channels, base_dir, cell_som_cluster_cols, max_
+    cell_som_cluster_avgs = pd.merge_asof(
+        cell_som_cluster_avgs, cell_cc.mapping, on='cell_som_cluster'
     )
 
     # resave average number of pixel/SOM clusters across all cell SOM clusters
     # with metacluster assignments
-    cell_som_cluster_avgs_and_counts.to_csv(
-        som_cluster_counts_avg_path,
+    cell_som_cluster_avgs.to_csv(
+        som_expr_col_avg_path,
         index=False
     )
 
-    # compute the weighted channel average expression per cell SOM cluster
+
+def generate_wc_avg_files(fovs, channels, base_dir, cell_cc,
+                          cell_som_input_data_name='cluster_counts_size_norm.feather',
+                          weighted_cell_channel_name='weighted_cell_channel.feather',
+                          cell_som_cluster_channel_avg_name='cell_som_cluster_channel_avg.csv',
+                          cell_meta_cluster_channel_avg_name='cell_meta_cluster_channel_avg.csv'):
+    """Generate the weighted channel average files per cell SOM and meta clusters.
+
+    Args:
+        fovs (list):
+            The list of fovs to subset on
+        channels (list):
+            The list of channels to subset on
+        base_dir (str):
+            The path to the data directory
+        cell_cc (cluster_helpers.PixieConsensusCluster):
+            The consensus cluster object containing the SOM to meta mapping
+        cell_som_input_data_name (str):
+            The input file used for SOM training. For weighted channel averaging, it should
+            contain the number of pixel SOM/meta cluster counts of each cell,
+            normalized by `cell_size`
+        weighted_cell_channel_name (str):
+            The name of the file containing the weighted channel expression table
+        cell_som_cluster_channel_avg_name (str):
+            The name of the file to save the average weighted channel expression
+            per cell SOM cluster
+        cell_meta_cluster_channel_avg_name (str):
+            Same as above except for cell meta clusters
+    """
+    # define the paths to the data
+    cell_som_input_data_path = os.path.join(base_dir, cell_som_input_data_name)
+    weighted_channel_path = os.path.join(base_dir, weighted_cell_channel_name)
+    som_cluster_channel_avg_path = os.path.join(base_dir, cell_som_cluster_channel_avg_name)
+    meta_cluster_channel_avg_path = os.path.join(base_dir, cell_meta_cluster_channel_avg_name)
+
+    # check paths
+    io_utils.validate_paths([cell_som_input_data_path, weighted_channel_path])
+
+    # if the weighted channel average files exist, skip
+    if os.path.exists(som_cluster_channel_avg_path) and \
+       os.path.exists(meta_cluster_channel_avg_path):
+        print("Already generated average weighted channel expression files, skipping")
+        return
+
     print("Compute average weighted channel expression across cell SOM clusters")
     cell_som_cluster_channel_avg = compute_cell_cluster_channel_avg(
```

Comment:
It seems to be that `compute_cell_cluster_channel_avg` and `compute_cell_cluster_expr_avg` are doing very similar things (and have very similar names) - any way to just combine those 2?

Reply:
We can spin up a combination solution, however there are three reasons I've left them separated:

1. Running `compute_cell_cluster_channel_avg` involves an additional file. This leads to enough components being different between the two (validation, subsetting, etc.) that I'd argue it's better to separate the two to avoid code bloat.
2. Code maintainability: if a future bug fix/patch/etc. involves the averaging step, it'd be (just a bit) easier to isolate the problem to one of these functions, as opposed to a single function that handles multiple cases.
3. Documentation: the way the functions are, it's clear what the column expression and channel averaging functions do and what arguments they take. I fear it'd be more difficult to understand if we combined all of this in one function.

@ngreenwald let me know what you think.

Reply:
I could see it going either way honestly, up to you guys

Reply:
I'm not going to die on this hill, but I think the functionality is so similar that it makes it more confusing for the user that they are two functions (especially with such similar names). Like when I was reviewing the code, I spent quite a bit of time trying to understand why it's 2 separate functions. In my view, code maintainability would actually be easier if it's one function since the functionality is basically the same. Basically all the function does is average over cell som/meta clusters for a set of input columns. The reason why `compute_cell_cluster_channel_avg` takes an additional input is because the clusters and the input features are in two different files - it's only a few lines to first combine those into one table, those lines could easily be moved to before wherever the function is being called.

But like I said, if you truly think it'll be annoying to combine them, we can leave it. But at the very lease, I think the function names should be changed so they're not so similar.

Reply:
@cliu72 how about we make a separate issue for this and revisit it later this month? I do agree it's worth exploring options here, but I think this is extending beyond the scope of this particular PR.

Reply:
At the very least, can we change the name of the functions as a part of this one? `compute_cell_cluster_channel_avg` and `compute_cell_cluster_expr_avg` have basically the same meaning.

Reply:
Done!